lib/plugins/aws/package/compile/functions/index.js

```diff
@@ -377,99 +378,121 @@ class AwsCompileFunctions {
         const versionHash = crypto.createHash('sha256');
         fileHash.setEncoding('base64');
         versionHash.setEncoding('base64');
+        const layerConfigurations = _.cloneDeep(
+          extractLayerConfigurationsFromFunction(functionResource.Properties, cfTemplate)
+        );
 
-        // Read the file in chunks and add them to the hash (saves memory and performance)
-        versionCompilationPromise = BbPromise.fromCallback(cb => {
-          const readStream = fs.createReadStream(artifactFilePath);
-
-          readStream
-            .on('data', chunk => {
-              fileHash.write(chunk);
-              versionHash.write(chunk);
-            })
-            .on('close', () => {
-              cb();
-            })
-            .on('error', error => {
-              cb(error);
+        // Include function and layer configuration in version id hash (without the Code part)
+        versionCompilationPromise = addFileContentsToHashes(artifactFilePath, [
+          fileHash,
+          versionHash,
+        ])
+          .then(() => {
+            // Include all referenced layer code in the version id hash
+            const layerArtifactPaths = [];
+            layerConfigurations.forEach(layer => {
+              const layerArtifactPath = this.provider.resolveLayerArtifactName(
+                layer.serverlessLayerName
+              );
+              layerArtifactPaths.push(layerArtifactPath);
             });
-        }).then(() => {
-          // Include function configuration in version id hash (without the Code part)
-          const properties = _.omit(_.get(functionResource, 'Properties', {}), 'Code');
-          _.forOwn(properties, value => {
-            if (_.isObject(value)) {
-              versionHash.write(JSON.stringify(value));
-            } else {
-              versionHash.write(value != null ? String(value) : '');
+
+            return layerArtifactPaths
+              .sort()
+              .reduce(
+                (previousHashCalculation, artifactName) =>
+                  previousHashCalculation.then(() =>
+                    addFileContentsToHashes(artifactName, [versionHash])
+                  ),
+                BbPromise.resolve()
+              );
+          })
+          .then(() => {
+            // Include function and layer configuration details in the version id hash
+            for (const layerConfig of layerConfigurations) {
+              delete layerConfig.properties.Content.S3Key;
             }
-          });
 
-          // Finalize hashes
-          fileHash.end();
-          versionHash.end();
+            // sort the layer conifigurations for hash consistency
+            const sortedLayerConfigurations = {};
+            const byKey = ([key1], [key2]) => key1.localeCompare(key2);
+            for (const { name, properties: layerProperties } of layerConfigurations) {
+              sortedLayerConfigurations[name] = _.fromPairs(_.entries(layerProperties).sort(byKey));
```

Comment:
Let's use `Object.entries` instead of `_.entries` (as  in a meantime we've dropped support for Node.js v8)